### PR TITLE
Add support for stacks

### LINKED ISSUE #137: Create a vetiver model for a stacked ensemble object
Hi Team,

Thanks for the great work you're doing. I have a slight issue which is detailed below;

* I used the stacks package to build an ensemble stacked ML following the tidy models approach. However, when I tried to deploy the model, I encountered an error while converting the stacked model object to the `vetiver_model` object.
* The screenshots below detail the object class and the error messages.

image

image

This is what my session info looks like;

image

### PR DESCRIPTION
**Pull request by juliasilge**
Closes #137 

``` r
library(vetiver)
library(stacks)

frog_reg <-
    stacks() %>%
    add_candidates(reg_res_lr) %>%
    add_candidates(reg_res_sp) %>%
    blend_predictions() %>%
    fit_members()

vetiver_model(frog_reg, "frog-stack")
#> 
#> ── frog-stack ─ <butchered_linear_stack> model for deployment 
#> A regression stacked ensemble with 4 members using 4 features
```

<sup>Created on 2022-08-25 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

### CODE REVIEW

**Comment by juliasilge** [comment]
In the course of doing this PR I was reminded (via `tree_frogs`) about how `vetiver_type_convert()` behaves for a character variable and have opened #139.

**Comment by juliasilge** [comment]
@simonpcouch would you mind doing a review of this PR, which adds stacks support to vetiver?

**Comment by simonpcouch** [comment]
@juliasilge I'm glad to! Will get to this on Monday. :)

File: R/stacks.R [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        nrow()
+    glue("A {model$mode} stacked ensemble with {num_members} members")
+}
+
+#' @rdname vetiver_create_meta
+#' @export
+vetiver_create_meta.model_stack <- function(model, metadata) {
+    reqs <- map(model$member_fits, required_pkgs)
+    reqs <- purrr::flatten_chr(reqs)
+    reqs <- sort(unique(c(reqs, "workflows", "stacks")))
```
Comment (simonpcouch) [inline R/stacks.R:16]: The `model$coefs` slot has a glmnet dependency, and may introduce others if we move forward with generalizing the meta-learner. Maybe:

```suggestion
    reqs <- sort(unique(c(reqs, required_pkgs(model$coefs), "workflows", "stacks")))
```

File: tests/testthat/test-stacks.R [hunk 2]
```diff
@@ -0,0 +1,71 @@
+skip_if_not_installed("stacks")
+skip_if_not_installed("ranger")
```
Comment (simonpcouch) [inline tests/testthat/test-stacks.R:2]: Unless this is a needed vetiver Suggests, we should be able to remove this line! The `class_res_rf` object is the stacks example object with a ranger dependency.

```suggestion

```

**Review by simonpcouch — COMMENTED** [review]
A few comments about oddities in the `model_stack` object structure, but otherwise this looks great! I enjoyed the opportunity to poke around in vetiver more thoroughly. :)

File: tests/testthat/_snaps/stacks.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
+# can print stacks model
+
+    Code
+      v
+    Output
+      
+      -- frog-stack - <butchered_linear_stack> model for deployment 
+      A regression stacked ensemble with 4 members using 4 features
```
[the remark this answers is not in the document]
Comment (juliasilge) [reply tests/testthat/_snaps/stacks.md]: Good idea -- I had to bump up to `penalty = 20` to keep less than four!

File: R/stacks.R [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    vetiver_meta(metadata, required_pkgs = reqs)
+}
+
+#' @rdname vetiver_create_ptype
+#' @export
+vetiver_ptype.model_stack <- function(model, ...) {
+    mold <- workflows::extract_mold(model$member_fits[[1]])
+    mold$blueprint$ptypes$predictors
```
[the remark this answers is not in the document]
Comment (juliasilge) [reply R/stacks.R:24]: Yes, that's right. The [mold](https://hardhat.tidymodels.org/articles/mold.html) is the record of the original input, not the result of applying preprocessing to the input:

``` r
library(tidymodels)
data("biomass")
rec <- recipe(HHV ~ carbon + hydrogen + oxygen + nitrogen + sulfur, 
              data = biomass)
mold1 <- workflow(rec, linear_reg()) %>% 
    fit(data = biomass) %>%
    extract_mold()
mold2 <- workflow(rec %>% step_pca(all_numeric_predictors()), linear_reg()) %>%
    fit(data = biomass) %>%
    extract_mold()

mold1$blueprint$ptypes$predictors
#> # A tibble: 0 × 5
#> # … with 5 variables: carbon <dbl>, hydrogen <dbl>, oxygen <dbl>,
#> #   nitrogen <dbl>, sulfur <dbl>
mold2$blueprint$ptypes$predictors
#> # A tibble: 0 × 5
#> # … with 5 variables: carbon <dbl>, hydrogen <dbl>, oxygen <dbl>,
#> #   nitrogen <dbl>, sulfur <dbl>
```

<sup>Created on 2022-08-31 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by juliasilge** [comment]
Thank you so much for the review @simonpcouch! 🙌

**Pull request merged**